### ISSUE TITLE
Add ignoreAtRules: [] to at-rule-name-newline-after and block-opening-brace-newline-before

### ISSUE THREAD

**Issue by Dromin** [stylelint/stylelint#2789]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request: currently there is an inconsistency in the available options of the group of rules of the types: `block-*-newline-*`, `block-*-space-*`, `block-*-empty-line-*` and similar `at-rule-*` rules. Some of these have an option `ignoreAtRules` which allows users to specify a list of at-rules to which the rule should not apply, while others do not have this option. I think the `ignoreAtRules` option should be available for all of these rules (for consistency, as well as to resolve a specific issue that I'm having).

Relevant `block-*` & `at-rule-*` rules that currently have an `ignoreAtRules` option:
-  block-closing-brace-newline-after
-  block-opening-brace-space-before
- at-rule-empty-line-before

Relevant `block-*` & `at-rule-*` rules that currently do not have an `ignoreAtRules` option:
- block-closing-brace-empty-line-before
-  block-closing-brace-newline-before
-  block-closing-brace-space-after
-  block-closing-brace-space-before
-  block-opening-brace-newline-after
-  block-opening-brace-newline-before
- block-opening-brace-space-after
- at-rule-name-newline-after
-  at-rule-name-space-after

> Which rule, if any, is this issue related to?

My specific issue is related to the rules `at-rule-name-newline-after` and `block-opening-brace-newline-before` (see the example below), but I believe all of the rules listed above need an `ignoreAtRules` option.

> What CSS is needed to reproduce this issue?

```scss
// This is fine; no rules are violated.
@mixin lorem-ipsum-im-a-mixin ($im-an-argument)
{
    color: yellowgreen;
}

// This violates `at-rule-name-newline-after` and `block-opening-brace-newline-before`.
@mixin im-another-mixin (
    $a-really-lengthy,
    $list-of,
    $arguments-which,
    $dont-fit,
    $on-a-single,
    $line
) {
    color: hotpink;
}
```

Explanation:

- `at-rule-name-newline-after`: the problem is related to preferred formatting differences between native CSS at-rules (e.g. `@media`) and SCSS-specific at-rules (e.g. `@mixin`, `@function`, etc.). I prefer to have the `at-rule-name-newline-after` rule set to `always-multi-line` for consistent formatting of multiline media queries, but in case of functions and mixins with multiline argument lists I'd like the function / mixin name to be on the same line as the at-rule. I wanted to solve this by setting this rule to ignore SCSS-specific at-rules, but this option is currently not available.

- `block-opening-brace-newline-before`: I set this to `always-multi-line` because I prefer to have an 'open' space between a selector and its declarations. However, this causes a problem in the case of functions / mixins with multiline argument lists; I'd like the closing bracket of the argument list and the opening bracket of the function / mixin block to be on the same line with a space in between (like in the example above) so that there is still a single line of 'open' space between the argument list and the declarations in the block. It's kind of a weird exception to the rule and I don't mind formatting this manually, but the problem is that, as with the other issue, I currently can't set this rule to ignore `@function` and `@mixin`.

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "block-opening-brace-newline-before": "always-multi-line",
    "at-rule-name-newline-after": "always-multi-line"
  }
}
```

> Which version of stylelint are you using?

`8.0.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Stylelint is integrated in my IDE (WebStorm).

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to SCSS-specific at-rules.

> What did you expect to happen?

The `ignoreAtRules` option should be available for the  `at-rule-name-newline-after` and `block-opening-brace-newline-before` rules, and ideally for other similar rules (listed above) as well.

> What actually happened (e.g. what warnings or errors you are getting)?

The `ignoreAtRules` option is currently not available for these rules.

**Comment by jeddy3** [comment]
@Dromin Thanks for the thorough suggestion and for using the issue template. Sorry about the delay triaging this issue (holidays and what not...).

> My specific issue is related to the rules `at-rule-name-newline-after` and `block-opening-brace-newline-before` (see the example below).

I agree. Adding an `ignoreAtRules: []` secondary to option to both these rules will resolve the issue for you.

> but I believe all of the rules listed above need an ignoreAtRules option.

As a rule of thumb we avoid adding features to stylelint until someone has a specific use case they have encountered. This is to reduce the overhead of reviewing and maintaining a feature that might never be used for anyone. I've a [pending PR](https://github.com/stylelint/stylelint/pull/2869) that documents this design decision.

I'm going to label this issue as "new option" and "help wanted" (and "beginner friendly" as it would make a good first pull request). Please consider contributing these options if you have time. There's [a section in the Developer Guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#adding-an-option-to-an-existing-rule) on how to get started.

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and add this option.
